### DISCUSSION TITLE
using-dependency-injection-in-a-net-core-console-application/

### DISCUSSION THREAD

**Discussion by andrewlock** [andrewlock/blog-comments#176] [Announcements]
# using-dependency-injection-in-a-net-core-console-application/

In this post, I show how to setup dependency injection in a .NET Core console app, both with the ASP.NET Core built-in container and with StructureMap.

https://andrewlock.net/using-dependency-injection-in-a-net-core-console-application/

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at December 27 2016, 08:16, in reply to [Fallon](https://disqus.com/by/disqus_p6hz8KUhiY/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/176#discussioncomment-6005616)</em>

---

<p>The majority of the examples I create are on GitHub at <a href="https://github.com/andrewlock/blog-examples" rel="nofollow noopener" title="https://github.com/andrewlock/blog-examples">https://github.com/andrewlock/blog-examples</a>, but I forgot to mention that in the post. You can find the example here: <a href="https://github.com/andrewlock/blog-examples/tree/master/using-dependency-injection-in-a-net-core-console-application" rel="nofollow noopener" title="https://github.com/andrewlock/blog-examples/tree/master/using-dependency-injection-in-a-net-core-console-application">https://github.com/andrewlock/blog-examples/tree/master/using-dependency-injection-in-a-net-core-console-application</a></p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at December 27 2016, 08:18, in reply to [Jan Strijk](https://disqus.com/by/janstrijk/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/176#discussioncomment-6005615)</em>

---

<p>Thanks! I've updated the post.</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at June 10 2017, 07:02, in reply to [Joep Beusenberg](https://disqus.com/by/joepbeusenberg/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/176#discussioncomment-6005624)</em>

---

<p>Not if you have a single class that wraps all the work that the console app does. You can use the provider to create the full web of dependencies then, and there's no need to pass the provider around.</p><p>If you don't have that single entry point, then yes, you would need to pass the provider around. </p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at August 07 2017, 05:25, in reply to [Joep Beusenberg](https://disqus.com/by/joepbeusenberg/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/176#discussioncomment-6005626)</em>

---

<p>Not really - everything always starts with a single class right? Program.Main. As long as you configure your service provider "high enough" up the call stack, then there shouldn't be any issues. But it's hard to discuss without an example of the problem you're having..</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at August 01 2017, 11:36, in reply to [Jay](https://disqus.com/by/disqus_2Dhhw1Vq1B/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/176#discussioncomment-6024578)</em>

---

<p>It depends how you configure StructureMap :) I suggest taking a look at the docs: <a href="http://structuremap.github.io/quickstart/" rel="nofollow noopener" title="http://structuremap.github.io/quickstart/">http://structuremap.github.io/quickstart/</a></p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at August 01 2017, 02:25, in reply to [Jay](https://disqus.com/by/disqus_2Dhhw1Vq1B/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/176#discussioncomment-6024580)</em>

---

<p>The point of this post was more about how to use DI in a .NET Core console app, rather than the benefit of StructureMap over the built in container. I have another post that touches on that a bit more here: <a href="https://andrewlock.net/getting-started-with-structuremap-in-asp-net-core/" rel="nofollow noopener" title="https://andrewlock.net/getting-started-with-structuremap-in-asp-net-core/">https://andrewlock.net/getting-started-with-structuremap-in-asp-net-core/</a></p><p>But in short - you don't have to configure everything manually with StructureMap, you generally just define conventions + you get nice diagnostics which are super useful when you have a complex project</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at March 09 2018, 05:34, in reply to [chris](https://disqus.com/by/disqus_Jfv14ng9wC/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/176#discussioncomment-6024586)</em>

---

<p>Yes, that works too</p>

[a comment by an automated account is not shown]

**Comment by alexeygritsenko** [comment]
Hi, is it correct that you don't use a scope in your example? Isn't creating a scope mandatory or isn't there some downside to not using one?
What if we replace this:
```
logger.LogDebug("Starting application");

//do the actual work here
var bar = serviceProvider.GetService<IBarService>();
bar.DoSomeRealWork();

logger.LogDebug("All done!");
```
With this:
```
logger.LogDebug("Starting application");

using (var scope = serviceProvider.CreateScope())
{
    //do the hard work here
    var bar = scope.ServiceProvider.GetService<IBarService>();
    bar.DoSomeRealWork();
}

logger.LogDebug("All done!");
```
